### ISSUE TITLE
Mocha parallel breaks pattern of shared before() and after()

### ISSUE THREAD

**Issue by Stono** [mochajs/mocha#4998]
### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [x] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend that you _not_ install Mocha globally.

### Description

Confusing title i know but hopefully the repro will help... we have a pattern where we have a common shared file that is imported from all our test suite that acts as a global `before` and `after` hook at the top level.  This works fine in a single thread, but when we use mocha parallel what we observe is the `after` being fired at the end of the workers first suite of tests, rather than at the end of them all.  It's easier to demonstrate.  The code sample below, run with jobs=1:

```
---> this is in global before undefined
  test 1
    ✔ undefined (1003ms)

  test 2
    ✔ undefined (1001ms)

  test 3
    ✔ undefined (1001ms)

---> this is in global after
```

As you can see, each test file is run, and the `before` and `after` fire where we'd expect them to.  `MOCHA_WORKER_ID` is `undefined` because we arent running in parallel. 

However run this with jobs=2, and you get:

```
---> this is in global before 0
---> this is in global before 1

  test 1
    ✔ 0 (1003ms)


  test 2
    ✔ 1 (1003ms)

---> this is in global after 0
---> this is in global after 1

  test 3
    ✔ 0 (1001ms)
```

Notice how the after hooks are firing after the first suite the worker runs.  It's more obvious when you have more test suite runs:

```
---> this is in global before 1
---> this is in global before 0

  test 2
    ✔ 1 (1002ms)


  test 1
    ✔ 0 (1001ms)

---> this is in global after 1
---> this is in global after 0

  test 3
    ✔ 1 (1001ms)


  test 4
    ✔ 0 (1001ms)


  test 5
    ✔ 1 (1002ms)


  test 6
    ✔ 0 (1002ms)
 ```
 
So notice here how `test 3` ran on `MOCHA_WORKER_ID` 0, but the `after` hook had already fired after `test 1` completed.  So it looks like the before and after tests in `common.ts` are only executing on the first suite that a worker runs, not the next suite.  

### Steps to Reproduce

`test1.test.ts`: 
```
import 'test/common'

describe('test 1', () => {
  it(`${process.env.MOCHA_WORKER_ID}`, (done) => {
    setTimeout(done, 1000)
  })
})
```

`test2.test.ts`: 
```
import 'test/common'

describe('test 2', () => {
  it(`${process.env.MOCHA_WORKER_ID}`, (done) => {
    setTimeout(done, 1000)
  })
})
```

`test3.test.ts`: 
```
import 'test/common'

describe('test 3', () => {
  it(`${process.env.MOCHA_WORKER_ID}`, (done) => {
    setTimeout(done, 1000)
  })
})
```

`common.ts`: 
```
before(() => {
  console.log('---> this is in global before', process.env.MOCHA_WORKER_ID)
})

after((done) => {
  done()
  setTimeout(() => {
    console.log('---> this is in global after', process.env.MOCHA_WORKER_ID)
  }, 50)
})
```

**Expected behavior:** [What you expect to happen]
Continuing the example above i'd expect to see the after hook either fire at the end of the worker, eg:

```
---> this is in global before 0
---> this is in global before 1

  test 1
    ✔ 0 (1003ms)


  test 2
    ✔ 1 (1003ms)

---> this is in global after 1

  test 3
    ✔ 0 (1001ms)
    
---> this is in global after 0
```

OR at the end of each suite a worker runs:

```
---> this is in global before 0
---> this is in global before 1

  test 1
    ✔ 0 (1003ms)


  test 2
    ✔ 1 (1003ms)

---> this is in global after 1
---> this is in global after 0
---> this is in global before 1
  test 3
    ✔ 0 (1001ms)
    
---> this is in global after 0
```

**Actual behavior:** [What actually happens]
See above

**Reproduces how often:** [What percentage of the time does it reproduce?]
100%

### Versions

- The output of `mocha --version` and `node_modules/.bin/mocha --version`: `10.2.0`
- The output of `node --version`: `v18.14.2`
- Your operating system
  - name and version: macosx
  - architecture (32 or 64-bit): arm64
- Your shell (e.g., bash, zsh, PowerShell, cmd): zsh
- Your browser and version (if running browser tests): n/a
- Any third-party Mocha-related modules (and their versions): n/a
- Any code transpiler (e.g., TypeScript, CoffeeScript, Babel) being used (and its version): typescript 5, ts-node

### Additional Information

.mocharc.json: 
```
{
  "require": [
    "should",
    "ts-node/register",
    "tsconfig-paths/register"
  ],
  "extensions": [
    "ts"
  ],
  "src": [
    "**/*.test.ts"
  ],
  "timeout": 60000
}
```

**Comment by Stono** [comment]
<img width="627" alt="Screenshot 2023-07-18 at 20 39 06" src="https://github.com/mochajs/mocha/assets/1486729/05ad14a7-9cfb-470e-9278-b6b4c0bb3678">

i think i've maybe just answered my issue in your docs... lol

**Comment by Stono** [comment]
Although saying that i suppose using your terminology i've defined a root hook, but that appears to being executed on the first suite on the worker?

**Comment by Stono** [comment]
So I had a suspicion that it was require cache.  You're spawning a nodejs process per worker but then executing suites across that worker.  The `import root hooks` isn't being evaluated properly on anything other than the first.  If you forcefully delete it from the require cache, eg:

```
delete require.cache[require.resolve('./common.ts')]
import 'test/common'
```

it works as i'd expect (before and after each suite):

```
---> this is in global before 0
---> this is in global before 1

  test 1
    ✔ 0 (1002ms)


  test 2
    ✔ 1 (1002ms)

---> this is in global before 0
---> this is in global before 1
---> this is in global after 0
---> this is in global after 1

  test 3
    ✔ 0 (1002ms)


  test 4
    ✔ 1 (1001ms)

---> this is in global before 0
---> this is in global before 1
---> this is in global after 0
---> this is in global after 1

  test 5
    ✔ 0 (1002ms)


  test 5
    ✔ 1 (1002ms)
```

**Comment by Stono** [comment]
FWIW i've fixed this in my space by moving to global hooks (beforeAll afterAll) injected via `require`.  Happy to close the issue but i still think the behaviour above is somewhat surprising and not what the documentation implies (it says to require in the top of each test file if you dont mind boilerplate).

**Comment by JoshuaKGoldberg** [comment]
> moving to global hooks (`beforeAll` `afterAll`) injected via `require`

Glad to see there's a solution! Even if it does involve working with `require`. 😄 

> Happy to close the issue but i still think the behaviour above is somewhat surprising and not what the documentation implies (it says to require in the top of each test file if you dont mind boilerplate).

This issue has a lot of context in it that's hard to read through. If you want to propose an update to docs that'd be great - but I'd request filing a new issue with a succinct proposal for what you think the new recommendation should be. For context, #5027: we're coming in fresh.

Thanks1
